### ISSUE TITLE
Could borg support reflink enabled filesystems?

### ISSUE THREAD

**Issue by alphazo** [borgbackup/borg#2743]
XFS has implemented some new (experimental) exciting features such as reflink that allows instant CoW snapshots similar to what is found on btrfs. I don't think there is a plan to support send/receive commands like on ZFS so the dedup function is pretty much limited to the local filesystem.
I'm envisioning to use the following scheme for my external USB drive that contains my photos and that I usually backup on a second drive or network storage using borg. This applies to both btrfs and the new reflink enabled xfs.

- While on the move and without network connectivity or access to my borg drive, I can, before working on the pictures, perform instant pseudo snapshots (not like the snapshots found on btrfs) without eating precious hard drive space.
`# cp -r --reflink=always /mnt/usbdrive/photos /mnt/usbdrive/snapshots/snap01`

- When I'm back home I can connect my photo drive and my backup drive to my PC and initiate a borg backup.
`# borg create /mnt/backupdrive/borg::borg-snap1 /mnt/usbdrive`

So the borg-snap1 snapshot will contain all the different snapshots I performed while away from home plus the working directory. But since borg doesn't know about the reflink feature it will rescan each of the files found in each snapshot found on my photo drive thinking they are new files but will ultimately find corresponding known dedup blocks so it will effectively not copy over each of the btrfs/xfs pseudo snapshot. I tried it and the size of my photo directore + many snapshots of the same pictures gave pretty much the size of the photo directory on the borg snapshot which is great. 
I was wondering it there would be a way to have a new feature in borg to detect such reflink enabled filesystems (btrfs/xfs) so it would immediately know that a file found in a btrfs/xfs directory is a duplicate of an existing known one and therefore use the same dedup blocks.

**Comment by enkore** [comment]
According to [1] there is no way to find physical extents (the backing element of reflinks) without either risking data corruption (when btrfs compression is used) or writing code that parses btrfs data structures. Apart from that it could be incorporated into the files cache (key := id-hash(NUL || "physical-extents" || extent-descriptor...)).

[1] https://www.spinics.net/lists/linux-btrfs/msg60845.html

**Comment by alphazo** [comment]
Would the new reverse mapping (rmapbt) support on xfs be of any help for identifying CoW files ?

https://lwn.net/Articles/695290/
https://lwn.net/Articles/659677/

**Comment by enkore** [comment]
I don't see a problem on XFS, apart from this being a rather fickle business. It's the btrfs issue I linked to above that seems problematic to me (there needs to be a reliable way to detect compressed files/extents to work around it). coreutils hints at problems with ext4, though the comments are old. Maybepossibly fixed.

**Comment by alphazo** [comment]
Understood. Thanks for takin the time to answer. Those new XFS features (reflink + rmapbt) are still marked as Experimental anyway. I'm going to give them a try on a controlled environment and see how they perform. By the time XFS reflink goes primetime more people might express a need for such feature on borg. I find it a good balance between btrfs and its flacky RAID support and ZFS that is not available straight in the kernel. Cheap snapshots + borg for real dedup backup is probably going to be my next workhorse.

**Comment by alphazo** [comment]
Some more pointers from the xfs folks:

> borgbackup will probably need to call the GETFSMAP ioctl, which won't land until 4.12.  On xfs, rmapbt is needed to supply data block ownership info, which is what borgbackup (and bees, and...) say they want to be smarter about dedup.


https://www.spinics.net/lists/linux-xfs/msg08128.html

**Comment by jcharaoui** [comment]
With the release of Linux 4.16, the XFS reflink feature [is no longer tagged EXPERIMENTAL](http://lkml.iu.edu/hypermail/linux/kernel/1801.3/05584.html).

**Comment by alphazo** [comment]
@jcharaoui Thanks for pointing this out. I used those features on nearly a year on my photo hard drive and haven't seen any problem. Now I believe that the rmapt feature is also no longer experimental (I'm running Linux 4.16.9) since I no longer see those red warnings when mounting my external drive that has both reflink and rmapbt enabled.

**Comment by srd424** [comment]
Interested in this while watching the first borg backup of a btrfs-based container pool take _forever_ :)
Note that [duperemove](https://github.com/markfasheh/duperemove) claims to check shared extents when working out whether to hash files ..

**Comment by charles-dyfis-net** [comment]
https://gist.github.com/charles-dyfis-net/bfb0e30862f04957d020afe0ff8b093b may be of interest to those here -- invoking `xfs_io` to reflink together identical chunks of content Borg has identified.

Not maintained, not recently tested, not documented at time of development and use, very much YMMV.

**Comment by srd424** [comment]
For my use case I'm now investigating https://github.com/systemd/casync, which has btrfs reflink support (don't know if it works on xfs.) I hit a few bugs, but worked out fixes for a couple (systemd/casync#235, systemd/casync#237 - now merged) and found work-arounds for the other two (systemd/casync#239, systemd/casync#240.)

**Comment by charles-dyfis-net** [comment]
> For my use case I'm now investigating https://github.com/systemd/casync, which has btrfs reflink support (don't know if it works on xfs.) I hit a few bugs, but worked out fixes for a couple ([systemd/casync#235](https://github.com/systemd/casync/issues/235), [systemd/casync#237](https://github.com/systemd/casync/issues/237) - now merged) and found work-arounds for the other two ([systemd/casync#239](https://github.com/systemd/casync/issues/239), [systemd/casync#240](https://github.com/systemd/casync/issues/240).)

At the risk of plugging a project I'm a contributor to, I strongly suggest also looking into https://github.com/folbricht/desync. casync may have gotten better over time, but back when desync was started, casync's error handling was atrocious.

**Comment by srd424** [comment]
I'd looked at desync and thought it didn't support reflinking, but it seems I might be mistaken .. will revisit! casync .. does have some issues.

**Comment by charles-dyfis-net** [comment]
> I'd looked at desync and thought it didn't support reflinking, but it seems I might be mistaken .. will revisit! casync .. does have some issues.

Depending on when you looked it may not have; but it most definitely does today. See https://github.com/folbricht/desync/blob/4a8700c059471d5f005dd7c9a957072bb1fa5c8a/fileseed.go#L123-L126

**Comment by srd424** [comment]
Yup, just found that - I only looked the other day but I think I'd mis-parsed the beginning of the README. Going off topic here, but quickly - desync doesn't seem to support ACLs in the catar stream, but does now implement xattrs - does that mean it can restore ACLs from catar streams that it generates itself?

**Comment by charles-dyfis-net** [comment]
> Yup, just found that - I only looked the other day but I think I'd mis-parsed the beginning of the README. Going off topic here, but quickly - desync doesn't seem to support ACLs in the catar stream, but does now implement xattrs - does that mean it can restore ACLs from catar streams that it generates itself?

Couldn't say. I'm a regular user and occasional contributor, but I don't use that particular functionality. Frank does have a gitter chat room, though -- I'd suggest asking in there.

**Comment by jackwasey** [comment]
multiple filesystems now support reflinks, (zfs, bcachefs, btrfs, xfs). it seems possible to test whether a whole file is a share the same extents as another file, e.g. `rmlint --is-reflink`. this would at the very least avoid scanning files to check for duplication if they are already known to be reflinked.

**Comment by ThomasWaldmann** [comment]
@jackwasey Thanks for the hint.

I am not familiar with reflinks, but what we need to create an archive item is the correct chunks list (chunkid, size) tuples.

Now can can get that list in 2 ways:
- either we can determine a file is unchanged (using borg's files cache): same ctime, same inode, same size - then we use the chunks list from the cache.
- or we read the file, run it through chunker and hasher and get the chunks list that way (and store any missing chunks into the repo)

Not sure how a similar optimization could be done using reflinks?

**Comment by enkore** [comment]
See comment #2, in principle it's simple, using a hash of the extents you get from [fiemap](https://docs.kernel.org/filesystems/fiemap.html) instead of inode attrs as the cache key. In practice I still think almost ten years later doing this might have very interesting edge cases.

**Comment by jackwasey** [comment]
I'm a mere user, so cannot offer any technical insight, but thanks for thinking about this. It just seemed to me that there exists a filesystem level to determine identity of files or chunks of files, deeper than hard-links, and that could give significant performance improvements.
